### PR TITLE
Fix issue #148 + refactor load_checkpoint.py

### PR DESCRIPTION
**Pull request by nzmora**
The root-cause of this issue is that DataParallel modules cannot execute on the CPU,
on machines that have both CPUs and GPUs.
Therefore, we don’t use DataParallel for models loaded for the CPUs, but we do wrap
the models with DataParallel when loaded on the GPUs (to make them run faster).
The names of module keys saved in a checkpoint file depend if the modules are wrapped
by a DataParallel module or not.  So loading a checkpoint that ran on the GPU onto a
CPU-model (and vice-versa) will fail on the keys.
This is all PyTorch and despite the community asking for a fix -
e.g. pytorch/pytorch#7457 - it is still pending.

This commit contains code to catch key errors when loading a GPU-generated model
(i.e. with DataParallel) onto a CPU, and convert the names of the keys.

This PR also merges refactoring to load_chackpoint.py done by @barrh, who also added
a test to further test loading checkpoints.

### CODE REVIEW

File: tests/test_infra.py [hunk 1]
```diff
@@ -34,7 +39,41 @@ def test_load():
...
+def test_load_state_dict():
+    # prepare lean checkpoint
+    state_dict_arrays = torch.load('../examples/ssl/checkpoints/checkpoint_trained_dense.pth.tar').get('state_dict')
+
+    with tempfile.NamedTemporaryFile() as tmpfile:
+        torch.save({'state_dict': state_dict_arrays}, tmpfile.name)
+        model = create_model(False, 'cifar10', 'resnet20_cifar')
+        model, compression_scheduler, start_epoch = load_checkpoint(model, tmpfile.name)
+
+    assert len(list(model.named_modules())) >= len([x for x in state_dict_arrays if x.endswith('weight')]) > 0
```
Comment (guyjacob) [inline tests/test_infra.py]: Not sure I understand what this is testing. Loading the checkpoint doesn't add or remove modules from the model. So this assertion will return the same result before and after loading the checkpoint.
Why not compare `model.state_dict()` to `state_dict_arrays` after the load? They should be identical.

**Comment by barrh** [comment]
I feel like this PR should be much cleaner, had it not been for the merge commits.
It's makes it difficult to review your proposal due to those merge commits.

I made up a clean branch: https://github.com/barrh/distiller/commits/load_cp

File: distiller/scheduler.py [hunk 2]
```diff
@@ -187,26 +189,31 @@ def state_dict(self):
         state = {'masks_dict': masks}
         return state
 
-    def load_state_dict(self, state):
+    def load_state_dict(self, state, convert_scheduler_keys):
```
[the remark this answers is not in the document]
Comment (nzmora) [reply distiller/scheduler.py]: good idea!

File: tests/test_infra.py [hunk 3]
```diff
@@ -34,7 +39,41 @@ def test_load():
...
         model = create_model(False, 'cifar10', 'resnet20_cifar')
         model, compression_scheduler, start_epoch = load_checkpoint(model, 'THIS_IS_AN_ERROR/checkpoint_trained_dense.pth.tar')
+
+
+def test_load_gpu_model_on_cpu():
+    model = create_model(False, 'cifar10', 'resnet20_cifar', device_ids=-1)
+    model, compression_scheduler, start_epoch = load_checkpoint(model,
+                                                                '../examples/ssl/checkpoints/checkpoint_trained_dense.pth.tar')
+    assert compression_scheduler is not None
+    assert start_epoch == 180
```
[the remark this answers is not in the document]
Comment (nzmora) [reply tests/test_infra.py]: you are right - I'll add a test for the placement of the tensors on the CPU.
